### ISSUE TITLE
WAIC calculation

### ISSUE THREAD

**Issue by segasai** [JohannesBuchner/ic4stan#1]
Hi Johannes, 

I was looking at your WAIC code, and are you sure it is correct ?  As far as I can see WAIC requires elppd (i.e. pointwise likelihood estimation) rather then elpd. E.g. I'm looking at Eq. 7.5 from Gelman's BDA3, and it implies that you need to not just sum the log-likelihoods over the chain, but you need to do sum_i of log (<likelihood(Data_i)>) where average should be taken over the posterior. I.e. to compute WAIC you need per point evaluation of the (likelihood). 
See also table 7.1 of BDA3. 

Am I wrong somewhere ? 

  Sergey

**Comment by JohannesBuchner** [comment]
Hi Sergey,

I am not 100% sure, but I took the implementation from either PyMC or loo. Would be glad if you could double-check.

Here are some resources:

LOO:
https://github.com/stan-dev/loo/blob/1c27fd4a59c9534d272d1f929d32eecb870144f7/R/helpers.R
https://github.com/stan-dev/loo/blob/1c27fd4a59c9534d272d1f929d32eecb870144f7/R/waic.R

PyMC:
https://stats.stackexchange.com/questions/173128/watanabe-akaike-widely-applicable-information-criterion-waic-using-pymc
https://github.com/pymc-devs/pymc3/issues/888
https://github.com/pymc-devs/pymc3/blob/f1e656cfc74cceb2d880dbdb8dc29ff7a7229acb/pymc3/stats.py

It also would be nice to have WBIC but it seems incalculable? At least I did not find any implementations.

**Comment by segasai** [comment]
https://stats.stackexchange.com/questions/173128/watanabe-akaike-widely-applicable-information-criterion-waic-using-pymc -- this reference matches my understanding of WAIC (and my own implementation of it). I also checked this https://github.com/stan-dev/loo/blob/1c27fd4a59c9534d272d1f929d32eecb870144f7/R/helpers.R#L34 and I think it also does exactly what BDA says (i.e. computing things pointwise). So I think your WAIC is not quite right indeed. 

Here is how my WAIC calculation looks like (chain here is a 2D array of samples, like -- likelihood function returning per-point likelihoods, x - input data):

```
def waic(x, chain, like):    
    ndat = len(x)
    nsamp = chain.shape[0]
    r = np.zeros((nsamp, ndat))
    for i in range(nsamp):
        l = like(chain[i], x, sum=False)
        r[i, :] = l

    lppd = (scipy.misc.logsumexp(r, axis=0) - np.log(nuse)).sum()
    # average log(likelihood) per point

    pwaic2 = np.var(r, ddof=1, axis=0).sum()
    # the correction 
    clppd = lppd - pwaic2
    return -2*clppd

```

**Comment by JohannesBuchner** [comment]
By pointwise you mean per data point? Then it's not possible to compute from Stan output directly?

**Comment by segasai** [comment]
Yes, per data point. And I don't think there is a way of computing WAIC without modifying the code of the Stan model. I guess somehow it may be possible within Stan to do the WAIC calculation by accumulating the per point likelihoods with some tricks, but myself I don't know how to do that.

**Comment by segasai** [comment]
You can actually check this www.stat.columbia.edu/~gelman/research/unpublished/loo_stan.pdf page 24 -- there they save in generated quantities section the per point log(likelihoods) that can then be used for WAIC, but it has to be per point!
